### PR TITLE
Queue proof clean up. (partial)

### PR DESCRIPTION
**Pull request by yuhui-zheng**
Description
-----------
Clean up proofs for:
- QueuePeek
- QueueReceive
- QueueReceiveFromISR
- QueueSemaphoreTake

Clean up initialization function:
- xUnconstrainedMutex
- xUnconstrainedQueueBoundedItemSize

Need to clean up initialization a little bit more. For now, just to get things going. 

Test Steps
-----------
This is a draft PR which is trying to use CI to validate nothing's broken. Once CI finishes, will double check states are not reduced due to this PR. 

Please don't review yet. 

Related Issue
-----------
N/A

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by yuhui-zheng** [comment]
I've compared test job running locally without this PR (before) and #101 (after). States in cbmc.txt before/after clean-up for these four proofs:

QueuePeek before: 481, after: 482.
QueueReceive before: 492, after: 500.
QueueReceiveFromISR before: 420, after: 426.
QueueSemaphoreTake before: 825, after: 826.

Posting for reference, not as a metric to accept clean-ups. But since we cleaned up initialization and and duplicated assumes, this increase somewhat makes sense.

File: FreeRTOS/Test/CBMC/include/queue_init.h [hunk 1]
```diff
@@ -122,20 +124,22 @@ QueueHandle_t xUnconstrainedQueue( void ) {
...
+	__CPROVER_assume(xQueue);
+
+	xQueue->cTxLock = nondet_int8_t();
+	xQueue->cRxLock = nondet_int8_t();
+	xQueue->uxLength = nondet_UBaseType_t();
+	xQueue->uxMessagesWaiting = nondet_UBaseType_t();
+
+	/* This is an invariant checked with a couple of asserts in the code base.
+	    If it is false from the beginning, the CBMC proofs are not able to succeed*/
+	__CPROVER_assume(xQueue->uxMessagesWaiting < xQueue->uxLength);
```
Comment (yuhui-zheng) [inline FreeRTOS/Test/CBMC/include/queue_init.h:137]: Note for Mark -- before we are comparing bounded variable with unbounded. Before this change, in some harness stubs, we are bounding ```xQueue->uxMessagesWaiting``` again due to this.

File: FreeRTOS/Test/CBMC/proofs/Queue/QueueSemaphoreTake/QueueSemaphoreTake_harness.c [hunk 2]
```diff
@@ -32,57 +32,26 @@
...
-		xQueueSemaphoreTake(xQueue, xTicksToWait);
-	}
+	__CPROVER_assume( xState != taskSCHEDULER_SUSPENDED || xTicksToWait == 0 );
+
+	/* This is for loop unwinding. */
+	__CPROVER_assume( xQueue->cTxLock = LOCK_BOUND - 1 );
+	__CPROVER_assume( xQueue->cRxLock = LOCK_BOUND - 1 );
+
+	/* Volatile member. */
+	((&(xQueue->xTasksWaitingToReceive))->xListEnd).pxNext->xItemValue = nondet_ticktype();
```
Comment (yuhui-zheng) [inline FreeRTOS/Test/CBMC/proofs/Queue/QueueSemaphoreTake/QueueSemaphoreTake_harness.c:47]: Note for Mark -- this is a todo for me. Will come back to this in the next round.

File: FreeRTOS/Test/CBMC/proofs/Queue/QueueReceive/QueueReceive_harness.c [hunk 3]
```diff
@@ -60,29 +60,19 @@ QueueHandle_t xQueue;
...
-    	if(!pvBuffer){
-    		xQueue->uxItemSize = 0;
-    	}
-    	xQueueReceive( xQueue, pvBuffer, xTicksToWait );
-    }
+	__CPROVER_assume( xState != taskSCHEDULER_SUSPENDED || xTicksToWait == 0 );
 
+	/* This is for loop unwinding. */
+	__CPROVER_assume( xQueue->cTxLock = LOCK_BOUND - 1 );
+	__CPROVER_assume( xQueue->cRxLock = LOCK_BOUND - 1 );
```
Comment (AniruddhaKanhere) [inline FreeRTOS/Test/CBMC/proofs/Queue/QueueReceive/QueueReceive_harness.c:76]: For clarity, can we do `__CPROVER_assume( xQueue->cTxLock = ( LOCK_BOUND - 1 ) );` in lines 75-76

File: FreeRTOS/Test/CBMC/proofs/Queue/QueueSemaphoreTake/QueueSemaphoreTake_harness.c [hunk 4]
```diff
@@ -32,57 +32,26 @@
...
-		__CPROVER_assume( (
-		  ( UBaseType_t ) listGET_ITEM_VALUE_OF_HEAD_ENTRY( &( xQueue->xTasksWaitingToReceive ) )
-		   <= ( ( UBaseType_t ) configMAX_PRIORITIES)));
-		xQueueSemaphoreTake(xQueue, xTicksToWait);
-	}
+	__CPROVER_assume( xState != taskSCHEDULER_SUSPENDED || xTicksToWait == 0 );
+
+	/* This is for loop unwinding. */
+	__CPROVER_assume( xQueue->cTxLock = LOCK_BOUND - 1 );
+	__CPROVER_assume( xQueue->cRxLock = LOCK_BOUND - 1 );
```
Comment (AniruddhaKanhere) [inline FreeRTOS/Test/CBMC/proofs/Queue/QueueSemaphoreTake/QueueSemaphoreTake_harness.c:44]: For clarity can we do `__CPROVER_assume( xQueue->cTxLock = ( LOCK_BOUND - 1 ) );` across lines 43 - 44

File: FreeRTOS/Test/CBMC/include/queue_init.h [hunk 5]
```diff
@@ -69,21 +69,23 @@ QueueHandle_t xUnconstrainedQueueBoundedItemSize( UBaseType_t uxItemSizeBound )
 	__CPROVER_assume(uxQueueStorageSize < CBMC_OBJECT_MAX_SIZE);
 	__CPROVER_assume(uxItemSize < uxQueueStorageSize/uxQueueLength);
```
[the remark this answers is not in the document]
Comment (yuhui-zheng) [reply FreeRTOS/Test/CBMC/include/queue_init.h:70]: For the first two comments -- good catch, thanks. 

For ```uxItemSize``` and ```uxQueueLength``` -- line 64 to line 70 will be changed to:
```
__CPROVER_assume(uxItemSize < uxItemSizeBound);

/* Hitting assertion in QueueGenericCreate is not an interesting case for memory safefy proof. */
__CPROVER_assume(uxQueueLength > 0);

/* QueueGenericCreate method does not check for multiplication overflow. */
__CPROVER_assume(uxQueueLength < (CBMC_OBJECT_MAX_SIZE/uxItemSize));
```

[the remark this answers is not in the document]
Comment (yuhui-zheng) [reply FreeRTOS/Test/CBMC/include/queue_init.h:70]: For the third comment -- I think the comment at line 59 only meant to say "why we are only covering a certain combinations of ```uxQueueLength * uxItemSize```. The alternative could be this function doesn't take any argument, and ```uxQueueLength > 0```  ```uxItemSize >= 0``` additionally with the constraint ```uxQueueLength * uxItemSize < CBMC_OBJECT_MAX_SIZE```. 

I'm running the alternative, and it's really slow... I'll wait for one run to finish, but likely we are not discovering new state. We may need a better wording at line 59 it seems.

File: FreeRTOS/Test/CBMC/include/queue_init.h [hunk 6]
```diff
@@ -69,21 +69,23 @@ QueueHandle_t xUnconstrainedQueueBoundedItemSize( UBaseType_t uxItemSizeBound )
...
-		   If it is false from the beginning, the CBMC proofs are not able to succeed*/
-		__CPROVER_assume(xQueue->uxMessagesWaiting < xQueue->uxLength);
-		xQueue->xTasksWaitingToReceive.uxNumberOfItems = nondet_UBaseType_t();
-		xQueue->xTasksWaitingToSend.uxNumberOfItems = nondet_UBaseType_t();
-		#if( configUSE_QUEUE_SETS == 1)
-			xQueueAddToSet(xQueue, xUnconstrainedQueueSet());
-		#endif
-	}
+	QueueHandle_t xQueue = xQueueGenericCreate(uxQueueLength, uxItemSize, ucQueueType);
+	__CPROVER_assume(xQueue);
```
[the remark this answers is not in the document]
Comment (yuhui-zheng) [reply FreeRTOS/Test/CBMC/include/queue_init.h:73]: Coincidentally, we had a similar discussion earlier today. Two types of harnesses calling ```xUnconstrainedQueueBoundedItemSize()``` as a step for setting up test -- private functions and public APIs. 

All public APIs will ```configASSERT``` on null queue handle, and in my opinion that may be an uninteresting test case for memory safety proofs. (We may need separate API test cases to confirm asserts always get hit. @alfred2g @cobusve ) As for private functions calling this init step, e.g. ```prvCopyDataToQueue_harness.c```, those don't usually check for NULLs. 

I do agree with you "Would this assumption be better in the proof harness (where all the assumptions should be listed if possible)?". How about making two ```xUnconstrainedQueueBoundedItemSize()```? For public APIs, proposing leave it this way. For private function, I could update them to call into the new initialization step to additionally check NULLs.

File: FreeRTOS/Test/CBMC/include/queue_init.h [hunk 7]
```diff
@@ -69,21 +69,23 @@ QueueHandle_t xUnconstrainedQueueBoundedItemSize( UBaseType_t uxItemSizeBound )
...
+	
+	xQueue->cTxLock = nondet_int8_t();
+	xQueue->cRxLock = nondet_int8_t();
+	xQueue->uxLength = nondet_UBaseType_t();
+	xQueue->uxMessagesWaiting = nondet_UBaseType_t();
+
+	/* This is an invariant checked with a couple of asserts in the code base.
+	If it is false from the beginning, the CBMC proofs are not able to succeed*/
+	__CPROVER_assume(xQueue->uxMessagesWaiting < xQueue->uxLength);
```
[the remark this answers is not in the document]
Comment (yuhui-zheng) [reply FreeRTOS/Test/CBMC/include/queue_init.h:82]: (Folding this comment. Will address with similar ones altogether.)

File: FreeRTOS/Test/CBMC/include/queue_init.h [hunk 8]
```diff
@@ -122,20 +124,22 @@ QueueHandle_t xUnconstrainedQueue( void ) {
...
-		   If it is false from the beginning, the CBMC proofs are not able to succeed*/
-		__CPROVER_assume(xQueue->uxMessagesWaiting < xQueue->uxLength);
-		xQueue->xTasksWaitingToReceive.uxNumberOfItems = nondet_UBaseType_t();
-		xQueue->xTasksWaitingToSend.uxNumberOfItems = nondet_UBaseType_t();
-		#if( configUSE_QUEUE_SETS == 1)
-			xQueueAddToSet(xQueue, xUnconstrainedQueueSet());
-		#endif
-	}
+	QueueHandle_t xQueue = xQueueCreateMutex(ucQueueType);
+	__CPROVER_assume(xQueue);
```
[the remark this answers is not in the document]
Comment (yuhui-zheng) [reply FreeRTOS/Test/CBMC/include/queue_init.h:128]: (Folding this comment. Will address with similar ones altogether.)

File: FreeRTOS/Test/CBMC/include/queue_init.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuhui-zheng) [reply FreeRTOS/Test/CBMC/include/queue_init.h:137]: (Folding this comment. Will address with similar ones altogether.)

**Comment by yuhui-zheng** [comment]
I'm "archiving" this one. We'll resume this once a better test plan is in place. For now this works... Apologies @markrtuttle.

**Pull request closed without merging**